Create synthesis vignette (#105)
terraref/tutorials: vignettes/04-synthesis-data.Rmd

The second analysis compares greenness from image data with canopy cover. 

## Get and join data

Here we combine two dataframes. 
The first contains all the canopy height values for 2017, which was created in the traits vignette. 
The second is the cumulative growing degree days for all of 2017, which were calculated from the daily minimum and maximum temperatures in the weather vignette. 
They are combined by their common column, the date. 

```{r synth_setup}
library(dplyr)
library(ggplot2)
library(jsonlite)
library(lubridate)
library(traits)
library(inflection)

options(betydb_url = "https://terraref.ncsa.illinois.edu/bety/",
        betydb_api_version = 'v1', 
        betydb_key = '9999999999999999999999999999999999999999')
```

```{r get_trait_data, message = FALSE}
trait <- betydb_query(table     = "search", 
                      trait     = "canopy_height", 
                      date      = "~2017",
                      limit     =  "none")

> trait = trait %>% 

Reviewer: ```suggestion
trait <- trait %>% 
```

Author: Sorry, I never got in the habit of using the weird arrow thing because I learned Python at the same time as R. I'll make sure to go back and change all of those.

Reviewer: would the keyboard shortcut "option" + "-" help?

Author: I do use that shortcut, I just need to make it a habit.

We are using a logistic growth model here because it is appropriate for the shape of the GDD-cover relationship. 

```{r model_get_parameters}
single_cultivar <- trait_weather_df %>% 
  filter(cultivar == "PI656026")

cap     <- 150
initial <- 25
mean    <- single_cultivar$mean[67]
gdd_cum <- single_cultivar$gdd_cum[67]
rate    <- ((log((cap/mean) - 1)) - initial)/gdd_cum

model_single_cultivar <- nls(mean ~ cap / (1 + exp(initial + rate * gdd_cum)), 
                             start = list(cap = cap, initial = initial, rate = rate),
                             data = single_cultivar, trace = TRUE)

single_cultivar <- single_cultivar %>% 
  mutate(mean_predict = coef(model_single_cultivar)[1] / (1 + exp(coef(model_single_cultivar)[2] + coef(model_single_cultivar)[3] * gdd_cum)))
```

We then use the parameters from a single cultivar to run a model for each of the rest of the cultivars. 
These results are used to plot the model predictions, which are shown as an orange line. 
We also calculated the inflection point from each cultivar's model, which will be used in the following section. 

```{r model_all_cultivars}

> model_parameters <- trait_weather_df %>% 

Reviewer: you are correct - now that I look at it the for loop seems more straightforward, especially to those unfamiliar w/ the map function. A few thoughts:
1. to clarify you could 
  - feel free to return to the for loop or
  - add a few inline comments to explain what is happening 
1. eventually (not now) it would be better just to fit all data w/ random effects ... this is where I often leave it to the user

Author: I chose to revert this (22db799fdc2ddaf44e574f450f350d0fad4e1728). I'll go back to the nest/map method if I can figure out a more elegant way, I'm sure there is one.

For the first analysis, we want to figure out how the number of sufficiently warm days affects the amount of canopy cover at our site. We do this by combining the canopy cover data with the meteorological data on growing degree days, then modeling and plotting their relationship. We are specifically interested in figuring out when the increase in canopy cover starts to slow down in response to warm temperature days. 

The second analysis compares greenness from image data with canopy cover. 

## Get and join data

Here we combine two dataframes. 
The first contains all the canopy height values for 2017, which was created in the traits vignette. 
The second is the cumulative growing degree days for all of 2017, which were calculated from the daily minimum and maximum temperatures in the weather vignette. 
They are combined by their common column, the date. 

```{r synth_setup}
library(dplyr)
library(ggplot2)
library(jsonlite)
library(lubridate)
library(traits)
library(inflection)

options(betydb_url = "https://terraref.ncsa.illinois.edu/bety/",

>         betydb_api_version = 'v1', 

Reviewer: ```suggestion
        betydb_api_version = 'v1')
```

>         betydb_key = '9999999999999999999999999999999999999999')

Reviewer: for public data, can remove this API key
```suggestion
```

Reviewer: Odd ... can you confirm that you installed the traits package from terraref/traits and not from cran?

```

```{r get_trait_data, message = FALSE}
trait <- betydb_query(table     = "search", 
                      trait     = "canopy_height", 
                      date      = "~2017",
                      limit     =  "none")

trait = trait %>% 
  mutate(day = as.Date(raw_date))
```

```{r get_weather_data}
weather <- fromJSON('https://terraref.ncsa.illinois.edu/clowder/api/geostreams/datapoints?stream_id=46431&since=2017-01-01&until=2017-12-31', flatten = FALSE)

weather <- weather$properties %>% 
  mutate(time = ymd_hms(weather$end_time))

daily_values = weather %>% 
  mutate(day = as.Date(time), 
         air_temp_converted = air_temperature - 273.15) %>% 
  group_by(day) %>% 
  summarise(min_temp = min(air_temp_converted), 
            max_temp = max(air_temp_converted), 
            gdd = ifelse(sum(min_temp, max_temp) / 2 > 10, 
                         (max_temp + min_temp) / 2 - 10, 0))

daily_values <- daily_values %>% 
  mutate(gdd_cum = cumsum(gdd))
```

```{r combine_trait_weather}

> trait_weather_df = full_join(trait, daily_values, by = "day")

Reviewer: ```suggestion
trait_weather_df <- full_join(trait, daily_values, by = "day")
```

> trait_weather_df = trait_weather_df %>% 

Reviewer: ```suggestion
trait_weather_df <- trait_weather_df %>% 
```

for(each_cultivar in unique(trait_weather_df$cultivar)){
  each_cultivar_df <- filter(trait_weather_df, cultivar == each_cultivar)
  each_cultivar_model <- nls(mean ~ cap / (1 + exp(initial + rate * gdd_cum)), 
                            start = list(cap = 120, initial = 25, rate = rate),
                            data = each_cultivar_df)
  each_cultivar_df <- each_cultivar_df %>% 
      mutate(mean_predict = coef(each_cultivar_model)[1] / (1 + exp(coef(each_cultivar_model)[2] + coef(each_cultivar_model)[3] * gdd_cum)))
  all_cultivars <- rbind(each_cultivar_df, all_cultivars)
}

ggplot(all_cultivars) +
  geom_point(aes(x = gdd_cum, y = mean)) +
  geom_line(aes(x = gdd_cum, y = mean_predict), color = "orange") +
  facet_wrap(~cultivar, scales = "free_y") +
  labs(x = "Cumulative growing degree days", y = "Canopy Height")
```

## Create histogram of growth rate for all cultivars

The last thing that we are going to do is assess the difference in this relationship among the cultivars. 
We are going to use the inflection point from the logistic growth model, which indicates when canopy cover stops increasing as quickly with increasingly more warm days. 

> The `inflection` R package is used for the calculation. 

Reviewer: please use analytical solution https://stats.stackexchange.com/a/39262 instead of this package.

Author: See 9c16e81

ggplot(inflection_points, aes(x = inf_value)) +
  geom_histogram() +
  xlim(min(all_cultivars$gdd_cum), max(all_cultivars$gdd_cum)) +
  labs(x = "Inflection points", y = "Number")
```

## Get image data

For the second analysis, we are going to put together some trait data with RGB sensor data. 
These data are for May 2018, and are read in below. 
The image data are in the raster format. 

```{r get_image_data}
library(raster)
library(rgdal)
library(stringr)

# File downloaded from: https://terraref.ncsa.illinois.edu/clowder/datasets/5c507cb74f0c436195ba8eb4?space=5c50512a4f0c436195b9ad67
tif_url <- "vignettes/rgb_geotiff_L1_ua-mac_2018-05-04__13-04-09-888_left.tif"

rgb_raster_1      <- raster(tif_url, band = 1)
rgb_raster_1_crop <- crop(rgb_raster_1, extent(rgb_raster_1, 1500, 2500, 3000, 4000))
plot(rgb_raster_1_crop)
cellStats(rgb_raster_1_crop, stat = "mean")
hist(rgb_raster_1_crop)

rgb_raster_2      <- raster(tif_url, band = 2)
rgb_raster_2_crop <- crop(rgb_raster_2, extent(rgb_raster_2, 1500, 2500, 3000, 4000))
plot(rgb_raster_2_crop)
cellStats(rgb_raster_2_crop, stat = "mean")
hist(rgb_raster_2_crop)
```

```{r get_trait_data_2, message = FALSE}

> trait2 <- betydb_query(table     = "search", 

Reviewer: please use more informative object names, in this case, canopy_cover.

                       trait     = "canopy_cover", 

>                        date      = "~2018 May",

Reviewer: See if this makes sense (e.g. you get a good enough regression and there are enough overlapping dates w/ manual measurements and images available)

```suggestion
                       date                   = "~2017",
                       method_name  = 'Canopy cover estimation from ground-based conventional photographs'
```

and / or 

```r
                       trait                   = "canopy_height",
                       method_name = "Manual canopy height" 
```

                       limit     =  "none")

> trait2 <- trait2 %>% 

Reviewer: This didn't make it into [Hadley's style guide](http://adv-r.had.co.nz/Style.html) but I think it is a good idea to avoid overwriting objects, mostly because overwriting an object can lead to errors that are difficult to debug in interactive use.

Here you could just pipe the result of the query to the mutate function. In other cases (unless the object is v. big) you can increment the object name. I think that when the object is very big, it is okay to break the rule.

Reviewer 2: Created new ticket for this

  mutate(day = as.Date(raw_date))
```

## Calculate greenness metric for each image

> We then want a summary statistic for the entire image data. 

Reviewer: ```suggestion
Next we are going to compare the canopy cover measurements with a summary of the image. As an example, we will compute an index of greeness for each image.
```

> The chosen statistic here uses the red and green bands from the image raster. 

Reviewer: Please cite 
[Rasmussen et al 2016](https://github.com/terraref/tutorials/files/2829295/1-s2.0-S0034425701002899-main.pdf) or similar

```suggestion
The chosen statistic here is the  normalised green-red difference index, NGRDI=(R-G)/(R+G) (Rasmussen et al., 2016), which uses the red and green bands from the image raster:
```

If bookdown compilies w/ latex (low priority) you can set off the equation in a new line like  `$$\textrm{NGDRI}=\frac{R-G}{R+G}$$`

This metric is a way to characterize the greenness of the area where the data were taken from. 

```{r image_calculations}

> add_rasters <- rgb_raster_2_crop + rgb_raster_1_crop

Reviewer: it isn't clear here where / how the NGDRI equation is implemented - I assume that the `_1_` and `_2_` refer to the red and green layers? It would be nice to have object names like red and green (?) that represent the different layers and then have the equation like 

```r
ngdri <- (red - green) / (red + green)
```

  minX <- NA
  minY <- NA
  maxX <- NA
  maxY <- NA
  for (c in unique(bounds)){
    p = getLatLon(c)
    if (is.na(minX) || (minX > p[2]))
      minX = p[2]
    if (is.na(minY) || (minY > p[1]))
      minY = p[1]
    if (is.na(maxX) || (maxX < p[2]))
      maxX = p[2]
    if (is.na(maxY) || (maxY < p[1]))
      maxY = p[1]
  }
  
  return (c(minX, minY, maxX, maxY))
}

# Setting up our options
options(betydb_url = "https://terraref.ncsa.illinois.edu/bety/",
        betydb_api_version = 'v1',
        betydb_key = readLines('.betykey'))

# Makiong the query for our site
sites <- betydb_query(table     = "sites",  
                      sitename  = "MAC Field Scanner Season 6 Range 19 Column 1")

# Assigning the geometry of the site (GeoJSON format)
site.geom <- sites$geometry

# Stripping out the extra information to get to the points
complete_str <- str_match_all(site.geom, '(\\(\\(\\((.*)\\)\\)\\))')[[1]][, 3]
bounds <- strsplit(complete_str, ', ')[[1]]

# Getting the bounding box of the polygon
bounding_box = getBounds(bounds)
```

These are the names of the full field RGB data for the month of May.
We will be extracting our plot data from these files.

> A compressed file containing these images can be found on [Box]().

Author: "Box" will be the link to the tar containing the .tif files?

Reviewer 2: It's filled in now. A reference to google drive with ZIP file

     'fullfield_L1_ua-mac_2018-05-28_rgb_stereovis_ir_sensors_plots_sorghum6_shade_rgb_eastedge_mn_thumb.tif'
     )
```

We will loop through these images, extract our plot data, and calculate the "greeness" of each extract.
We are using the name of the file to extract the date for later.

```{r synth_get_greeness}
library(raster)
library(stringr)

# Extract the date from the file name
getDate <- function(file_name){
  date <- str_match_all(file_name, '[0-9]{4}-[0-9]{2}-[0-9]{2}')[[1]][,1]
  return(date)
}

# Get the clip coordinates into the correct order
clip_coords <- paste(toString(bounding_box[4]),' ',toString(bounding_box[3]),
                    ' ',toString(bounding_box[2]),' ',toString(bounding_box[1]))

# Returns the greeness value of the clipped image
getGreeness <- function(file_name, clip_coords){
  out_file <- "extract.tif"

  # Execute the GDAL command to extract the plot  
  command = paste("gdal_translate -projwin ",clip_coords," ",file_name," ",out_file)
  system(command)

  # Load the green band of the image and calculate the greeness value
  green_image   <- raster(out_file, band = 2)

>   greeness      <- cellStats(green_image, stat = "mean")

Author: Wasn't this supposed to be some summary metric that used both the green and red bands? Or did yall decide to change it to something simpler?

Reviewer: I think Chris may have missed discussion of the index ? Greenness alone is likely insufficient especially given the high variance in brightness across and even within images.
